baseapp/abci.go

```diff
@@ -360,6 +360,20 @@ func (app *BaseApp) Commit() (res abci.ResponseCommit) {
 	// MultiStore (app.cms) so when Commit() is called is persists those values.
 	app.deliverState.ms.Write()
 	commitID := app.cms.Commit()
+
+	res := abci.ResponseCommit{
+		Data:         commitID.Hash,
+		RetainHeight: retainHeight,
+	}
+
+	// call the hooks with the Commit message
+	for _, streamingListener := range app.abciListeners {
```

Comment:
I think this defers for different use cases, in my use case we prefer "sync" semantic for avoid losing messages.
Maybe we need to distinguish `asyncABCIListeners` and `syncABCIListeners`.

Reply:
On a second thought, async semantic can be implemented by the listener itself, but sync semantic can't if we don't call it synchronously in abci event.
the listener implementer should be aware of this, and we could provide some utilities if necessary.

Reply:
Yeah I understand the possibility of lost messages. I just worry this will impact `Commit()` time and I think that's really worth considering.

What about the panic?

Reply:
I don't think we should be calling `panic` under any circumstances for state streaming. Be it sync or async. I wouldn't want a node to halt due to a state listener being faulty IMO.

I personally would prefer to go with the async approach.

I would love others to chime in. @tac0turtle @aaronc @kocubinski @amaurym thoughts ?

Reply:
https://github.com/cosmos/cosmos-sdk/pull/13473 introduces a configuration option `stopNodeOnStreamingErr` which determines whether or not to panic on streaming error.  We ought to incorporate that here too, false by default.

Reply:
I think it's OK for the ABCI lifecycle methods to call into `abciListeners` synchronously.  Since `ABCIListener` is an interface leaving the choice of sync/async processing to the listener implementation seems sensible.  Hopefully the implementer is aware of the trade offs, especially in `ListenCommit`.

Reply:
> Since ABCIListener is an interface leaving the choice of sync/async processing to the listener implementation seems sensible.

In that case, we don't even need the `stopNodeOnStreamingErr`, since listener implementation can make the choice internally?

Reply:
> In the light of that, we don't even need the stopNodeOnStreamingErr, since listener implementation can make the choice internally?

Makes sense to me.  If this configuration option has a place, it's in the listener configurations ideally passed in during construction.